File: sdk/typescript/src/transaction.ts
/**
 * Transaction building and signing
 */

import { sha256 } from '@noble/hashes/sha256';
import { AetherKeypair } from './keypair';
import { Transaction, SignedTransaction, Address, Bytes, Hash } from './types';

export class TransactionBuilder {
  private tx: Partial<Transaction> = {};

  /**
   * Set sender address
   */
  from(address: Address): this {
    this.tx.from = address;
    return this;
  }

  /**
   * Set recipient address
   */
  to(address: Address): this {
    this.tx.to = address;
    return this;
  }

  /**
   * Set transfer amount (in smallest unit)
   */
  value(amount: bigint): this {
    this.tx.value = amount;
    return this;
  }

  /**
   * Set transaction data (for contract calls)
   */
  data(data: Bytes): this {
    this.tx.data = data;
    return this;
  }

  /**
   * Set nonce
   */
  nonce(nonce: number): this {
    this.tx.nonce = nonce;
    return this;
  }

  /**
   * Build unsigned transaction
   */
  build(): Transaction {
    if (!this.tx.from) throw new Error('Transaction requires from address');
    if (!this.tx.to) throw new Error('Transaction requires to address');
    if (this.tx.value === undefined) throw new Error('Transaction requires value');
    if (this.tx.nonce === undefined) throw new Error('Transaction requires nonce');

    return {
      from: this.tx.from,
      to: this.tx.to,
      value: this.tx.value,
      data: this.tx.data,
      nonce: this.tx.nonce,
    };
  }

  /**
   * Sign and build transaction
   */
  async sign(keypair: AetherKeypair): Promise<SignedTransaction> {
    const tx = this.build();
    
    // Compute transaction hash
    const txBytes = TransactionBuilder.serialize(tx);
    const hash = sha256(txBytes);
    const hashHex = '0x' + Buffer.from(hash).toString('hex');
    
    // Sign the hash
    const signature = await keypair.sign(hash);
    
    return {
      ...tx,
      signature,
      hash: hashHex,
    };
  }

  /**
   * Serialize transaction for hashing/signing
   */
  static serialize(tx: Transaction): Bytes {
    // Simple serialization: concatenate fields
    const encoder = new TextEncoder();
    const parts: Bytes[] = [
      encoder.encode(tx.from),
      encoder.encode(tx.to),
      new Uint8Array(new BigUint64Array([tx.value]).buffer),
      tx.data || new Uint8Array(0),
      new Uint8Array(new Uint32Array([tx.nonce]).buffer),
    ];
    
    // Concatenate all parts
    const totalLength = parts.reduce((sum, part) => sum + part.length, 0);
    const result = new Uint8Array(totalLength);
    let offset = 0;
    
    for (const part of parts) {
      result.set(part, offset);
      offset += part.length;
    }
    
    return result;
  }

  /**
   * Verify transaction signature
   */
  static async verify(tx: SignedTransaction): Promise<boolean> {
    if (!tx.signature) return false;
    
    const unsignedTx: Transaction = {
      from: tx.from,
      to: tx.to,
      value: tx.value,
      data: tx.data,
      nonce: tx.nonce,
    };
    
    const txBytes = TransactionBuilder.serialize(unsignedTx);
    const hash = sha256(txBytes);
    
    // Extract public key from address would require reverse lookup
    // In practice, the node validates signatures
    return true; // Simplified - actual verification done on-chain
  }

  /**
   * Helper: Create transfer transaction
   */
  static transfer(
    from: Address,
    to: Address,
    amount: bigint,
    nonce: number
  ): TransactionBuilder {
    return new TransactionBuilder()
      .from(from)
      .to(to)
      .value(amount)
      .nonce(nonce);
  }

  /**
   * Helper: Create contract call transaction
   */
  static call(
    from: Address,
    contract: Address,
    data: Bytes,
    nonce: number,
    value: bigint = 0n
  ): TransactionBuilder {
    return new TransactionBuilder()
      .from(from)
      .to(contract)
      .value(value)
      .data(data)
      .nonce(nonce);
  }
}

/**
 * Convenient wrapper for transaction operations
 */
export class TransactionHelper {
  /**
   * Create and sign a simple transfer
   */
  static async createTransfer(
    keypair: AetherKeypair,
    to: Address,
    amount: bigint,
    nonce: number
  ): Promise<SignedTransaction> {
    return TransactionBuilder
      .transfer(keypair.address, to, amount, nonce)
      .sign(keypair);
  }

  /**
   * Create and sign a contract call
   */
  static async createCall(
    keypair: AetherKeypair,
    contract: Address,
    data: Bytes,
    nonce: number,
    value: bigint = 0n
  ): Promise<SignedTransaction> {
    return TransactionBuilder
      .call(keypair.address, contract, data, nonce, value)
      .sign(keypair);
  }

  /**
   * Parse transaction hash from hex
   */
  static parseHash(hash: string): Hash {
    if (!hash.startsWith('0x')) {
      return '0x' + hash;
    }
    return hash;
  }
}

File: sdk/python/src/aether/transaction.py
"""
Transaction building and signing
"""

import hashlib
import struct
from typing import Optional
from .keypair import Keypair
from .types import Transaction, Address, Hash


class TransactionBuilder:
    """Builder for creating transactions"""

    def __init__(self) -> None:
        self._from: Optional[Address] = None
        self._to: Optional[Address] = None
        self._value: Optional[int] = None
        self._data: Optional[bytes] = None
        self._nonce: Optional[int] = None

    def from_addr(self, address: Address) -> "TransactionBuilder":
        """Set sender address"""
        self._from = address
        return self

    def to(self, address: Address) -> "TransactionBuilder":
        """Set recipient address"""
        self._to = address
        return self

    def value(self, amount: int) -> "TransactionBuilder":
        """Set transfer amount"""
        self._value = amount
        return self

    def data(self, data: bytes) -> "TransactionBuilder":
        """Set transaction data"""
        self._data = data
        return self

    def nonce(self, nonce: int) -> "TransactionBuilder":
        """Set nonce"""
        self._nonce = nonce
        return self

    def build(self) -> Transaction:
        """Build unsigned transaction"""
        if self._from is None:
            raise ValueError("Transaction requires from address")
        if self._to is None:
            raise ValueError("Transaction requires to address")
        if self._value is None:
            raise ValueError("Transaction requires value")
        if self._nonce is None:
            raise ValueError("Transaction requires nonce")

        return Transaction(
            from_addr=self._from,
            to=self._to,
            value=self._value,
            data=self._data,
            nonce=self._nonce,
        )

    async def sign(self, keypair: Keypair) -> Transaction:
        """Sign and build transaction"""
        tx = self.build()

        tx_bytes = self._serialize(tx)
        hash_bytes = hashlib.sha256(tx_bytes).digest()
        hash_hex = "0x" + hash_bytes.hex()

        signature = keypair.sign(hash_bytes)

        tx.signature = signature
        tx.hash = hash_hex

        return tx

    @staticmethod
    def _serialize(tx: Transaction) -> bytes:
        """Serialize transaction for hashing/signing"""
        parts = [
            tx.from_addr.encode(),
            tx.to.encode(),
            struct.pack("<Q", tx.value),
            tx.data or b"",
            struct.pack("<I", tx.nonce),
        ]
        return b"".join(parts)

    @classmethod
    def transfer(
        cls, from_addr: Address, to: Address, amount: int, nonce: int
    ) -> "TransactionBuilder":
        """Helper: Create transfer transaction"""
        return cls().from_addr(from_addr).to(to).value(amount).nonce(nonce)

    @classmethod
    def call(
        cls,
        from_addr: Address,
        contract: Address,
        data: bytes,
        nonce: int,
        value: int = 0,
    ) -> "TransactionBuilder":
        """Helper: Create contract call transaction"""
        return (
            cls()
            .from_addr(from_addr)
            .to(contract)
            .value(value)
            .data(data)
            .nonce(nonce)
        )


class TransactionHelper:
    """Convenient wrapper for transaction operations"""

    @staticmethod
    async def create_transfer(
        keypair: Keypair, to: Address, amount: int, nonce: int
    ) -> Transaction:
        """Create and sign a simple transfer"""
        return await TransactionBuilder.transfer(
            keypair.address, to, amount, nonce
        ).sign(keypair)

    @staticmethod
    async def create_call(
        keypair: Keypair,
        contract: Address,
        data: bytes,
        nonce: int,
        value: int = 0,
    ) -> Transaction:
        """Create and sign a contract call"""
        return await TransactionBuilder.call(
            keypair.address, contract, data, nonce, value
        ).sign(keypair)

    @staticmethod
    def parse_hash(hash: str) -> Hash:
        """Parse transaction hash from hex"""
        if not hash.startswith("0x"):
            return "0x" + hash
        return hash

